modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayCollector.groovy

```diff
@@ -0,0 +1,206 @@
+            sync.unlock()
+        }
+    }
+
+    /**
+     * Close the collector, submitting any remaining tasks as a partial job array.
+     */
+    void close() {
+        sync.lock()
+        try {
+            if( array.size() == 1 )
+                executor.submit(array.first())
+            else if( array.size() > 0 )
+                submit0(array)
+
+            closed = true
+        }
+        finally {
+            sync.unlock()
+        }
+    }
+
+    protected void submit0(List<TaskRun> tasks) {
+        executor.submit(createTaskArray(tasks))
+    }
+
+    /**
+     * Create the task run for a job array.
+     *
+     * @param tasks
+     */
+    protected TaskRun createTaskArray(List<TaskRun> tasks) {
+        // prepare child job launcher scripts
+        final handlers = tasks.collect( t -> executor.createTaskHandler(t) )
+        for( TaskHandler handler : handlers )
+            handler.prepareLauncher()
+
+        // create work directory
+        final hash = CacheHelper.hasher( tasks.collect( t -> t.getHash().asLong() ) ).hash()
+        final workDir = FileHelper.getWorkFolder(executor.getWorkDir(), hash)
+
+        Files.createDirectories(workDir)
+
+        // create wrapper script
+        final script = createTaskArrayScript(handlers)
+
+        // create config for job array
+        final rawConfig = new HashMap<String,Object>(ARRAY_DIRECTIVES.size())
+        for( final key : ARRAY_DIRECTIVES ) {
+            final value = processor.config.get(key)
+            if( value != null )
+                rawConfig[key] = value
+        }
+
+        // create job array
+        final first = tasks.min( t -> t.index )
+        final taskArray = new TaskArrayRun(
+            id: first.id,
+            index: first.index,
+            processor: processor,
+            type: processor.taskBody.type,
+            config: new TaskConfig(rawConfig),
+            context: new TaskContext(processor),
+            hash: hash,
+            workDir: workDir,
+            script: script,
+            children: handlers
+        )
+
+        return taskArray
+    }
+
+    /**
+     * Create the wrapper script for a job array.
+     *
+     * @param array
+     */
+    protected String createTaskArrayScript(List<TaskHandler> array) {
+        // get work directory and launch command for each task
+        final workDirs = array.collect( h -> h.getWorkDir() )
+        final args = array.first().getLaunchCommand().toArray() as String[]
+        final cmd = Escape.cli(args).replaceAll(workDirs.first(), '\\${task_dir}')
+
+        // create wrapper script
+        final indexName = executor.getArrayIndexName()
+        final indexStart = executor.getArrayIndexStart()
+        final arrayIndex = indexStart > 0
+            ? "${indexName} - ${indexStart}"
+            : "${indexName}"
+
+        """
+        array=( ${workDirs.collect( p -> Escape.path(p) ).join(' ')} )
+        export task_dir=\${array[${arrayIndex}]}
```

Comment:
it could be re-used `NXF_CHDIR` https://github.com/nextflow-io/nextflow/blob/49f7fbd144910f04e82d04909fb6a98e01fac542/modules/nextflow/src/main/resources/nextflow/executor/command-run.txt#L137-L137